pkg/util/parse/parse.go

```diff
@@ -0,0 +1,406 @@
+				return errors.New("propertyVisitor not found")
+			}
+		}
+
+		for k := 0; k < len(instance.Datas.Properties); k++ {
+			name := instance.Datas.Properties[k].PropertyName
+			l := 0
+			for l = 0; l < len(instance.PropertyVisitors); l++ {
+				if name == instance.PropertyVisitors[l].PropertyName {
+					instance.Datas.Properties[k].PVisitor = &instance.PropertyVisitors[l]
+					break
+				}
+			}
+			if l == len(instance.PropertyVisitors) {
+				return errors.New("propertyVisitor not found")
+			}
+		}
+
+		devices[instance.ID] = new(common.DeviceInstance)
+		devices[instance.ID] = &instance
+		klog.V(4).Info("Instance: ", instance.ID, instance)
+	}
+
+	for i := 0; i < len(deviceProfile.DeviceModels); i++ {
+		dms[deviceProfile.DeviceModels[i].Name] = deviceProfile.DeviceModels[i]
+	}
+
+	for i := 0; i < len(deviceProfile.Protocols); i++ {
+		protocols[deviceProfile.Protocols[i].Name] = deviceProfile.Protocols[i]
+	}
+	return nil
+}
+
+func ParseByUsingMetaServer(cfg *config.Config,
+	devices map[string]*common.DeviceInstance,
+	dms map[string]common.DeviceModel,
+	protocols map[string]common.Protocol) error {
+	// TODO it may be get all device from namespace
+	deviceList, err := httpclient.GetDeviceList(cfg.DevInit.MetaServer.Addr, cfg.DevInit.MetaServer.Namespace)
+	if err != nil {
+		return err
+	}
+	if len(deviceList) == 0 {
+		return ErrEmptyData
+	}
+	deviceModelList, err := httpclient.GetDeviceModelList(cfg.DevInit.MetaServer.Addr, cfg.DevInit.MetaServer.Namespace)
+	if err != nil {
+		return err
+	}
+	if len(deviceModelList) == 0 {
+		return ErrEmptyData
+	}
+	// TODO test only
+	//deviceList := tmpDevices
+	//deviceModelList := tmpDeviceModels
+	modelMap := make(map[string]common.DeviceModel)
+	for _, model := range deviceModelList {
+		cur := ParseDeviceModel(&model)
+		modelMap[model.Name] = cur
+	}
+
+	for _, device := range deviceList {
+		commonModel := modelMap[device.Spec.DeviceModelRef.Name]
+		protocol, err := BuildProtocol(&device)
+		if err != nil {
+			return err
+		}
+		instance, err := ParseDevice(&device, &commonModel)
+		if err != nil {
+			return err
+		}
+		instance.PProtocol = protocol
+		devices[instance.ID] = new(common.DeviceInstance)
+		devices[instance.ID] = instance
+		klog.V(4).Info("Instance: ", instance.ID)
+		dms[instance.Model] = modelMap[instance.Model]
+		protocols[instance.ProtocolName] = protocol
+	}
+
+	return nil
+}
+
+func ParseByUsingRegister(cfg *config.Config,
+	devices map[string]*common.DeviceInstance,
+	dms map[string]common.DeviceModel,
+	protocols map[string]common.Protocol) error {
+	klog.Infoln("======parse device and model from register")
+	deviceList, deviceModelList, err := grpcclient.RegisterMapper(cfg, true)
+	if err != nil {
+		return err
+	}
+
+	if len(deviceList) == 0 || len(deviceModelList) == 0 {
+		return ErrEmptyData
+	}
+	modelMap := make(map[string]common.DeviceModel)
+	for _, model := range deviceModelList {
+		klog.Infof("======ParseByUsingRegister model %+v", model)
```

Comment:
these log in this file are for test. please remove them.

Reply:
done